src/test/operators/join_test_runner.cpp

```diff
@@ -0,0 +1,527 @@
+#include <fstream>
+
+#include "base_test.hpp"
+#include "json.hpp"
+#include "operators/join_hash.hpp"
+#include "operators/join_index.hpp"
+#include "operators/join_mpsm.hpp"
+#include "operators/join_nested_loop.hpp"
+#include "operators/join_sort_merge.hpp"
+#include "operators/join_verification_operator.hpp"
+#include "operators/print.hpp"
+#include "operators/table_wrapper.hpp"
+#include "utils/load_table.hpp"
+#include "utils/make_bimap.hpp"
+
+/**
+ * This file contains the main tests for Hyrise's join operators.
+ * Testing is done by comparing the result of any given join operator with that of the JoinVerificationOperator for a
+ * number of configurations (i.e. JoinModes, predicates, data types, ...)
+ */
+
+using namespace std::string_literals;  // NOLINT
+
+namespace {
+
+using namespace opossum;  // NOLINT
+
+enum class InputSide { Left, Right };
+
+enum class InputTableType {
+  // Input Tables are data
+  Data,
+  // Input Tables are reference Tables with all Segments of a Chunk having the same PosList
+  SharedPosList,
+  // Input Tables are reference Tables with each Segment using a different PosList
+  IndividualPosLists
+};
+
+std::unordered_map<InputTableType, std::string> input_table_type_to_string{
+    {InputTableType::Data, "Data"},
+    {InputTableType::SharedPosList, "SharedPosList"},
+    {InputTableType::IndividualPosLists, "IndividualPosLists"}};
+
+struct InputTableConfiguration {
+  InputSide side{};
+  ChunkOffset chunk_size{};
+  size_t table_size{};
+  InputTableType table_type{};
+
+  auto to_tuple() const { return std::tie(side, chunk_size, table_size, table_type); }
+};
+
+bool operator<(const InputTableConfiguration& l, const InputTableConfiguration& r) {
+  return l.to_tuple() < r.to_tuple();
+}
+
+class BaseJoinOperatorFactory;
+
+struct JoinTestConfiguration {
+  InputTableConfiguration input_left;
+  InputTableConfiguration input_right;
+  JoinMode join_mode{JoinMode::Inner};
+  DataType data_type_left{DataType::Int};
```

Comment:
These defaults seem somewhat arbitrary. Would it be clearer if those parameters were explicitly set?

Reply:
When varying only a subset of the dimensions of the configuration, there have to be *some* defaults, we just don't get around that. Explicitly typing all defaults of in each place where a configuration is generated, is not better: It generates noise and makes it look like all those values were somehow chosen deliberately when they were actually not:

Compare this
```c++
    // Test all combinations of reference/data input tables. This tests mostly the composition of the output table
    for (const auto& left_input_table_type : all_input_table_types) {
      for (const auto& right_input_table_type : all_input_table_types) {
        auto join_test_configuration = default_configuration;
        join_test_configuration.input_left.table_type = left_input_table_type;
        join_test_configuration.input_right.table_type = right_input_table_type;

        add_configuration_if_supported(join_test_configuration);
      }
    }
```

to this
```c++
 // Test all combinations of reference/data input tables. This tests mostly the composition of the output table
for (const auto& left_input_table_type : all_input_table_types) {
  for (const auto& right_input_table_type : all_input_table_types) {
    auto join_test_configuration = JoinTestConfiguration;
    join_test_configuration.input_left.input_side = InputSide::Left;
    join_test_configuration.input_left.chunk_size = all_chunk_sizes.front();
    join_test_configuration.input_left.table_size = 10;
    join_test_configuration.input_left.table_type = left_input_table_type;
    join_test_configuration.input_left.input_side = InputSide::Right;
    join_test_configuration.input_left.chunk_size = all_chunk_sizes.front();
    join_test_configuration.input_left.table_size = 10;
    join_test_configuration.input_right.table_type = right_input_table_type;
    join_test_configuration.input_right.join_mode = JoinMode::Inner;
    join_test_configuration.input_right.data_type_left = DataType::Int;
    join_test_configuration.input_right.data_type_right = DataType::Int;
    join_test_configuration.input_right.nullable_left = true;
    join_test_configuration.input_right.nullable_right = true;
    join_test_configuration.predicate_condition = PredicateCondition::Equals;
    join_test_configuration.secondary_predicates = all_secondary_predicate_sets.front();
    join_test_configuration.operator_factory = std::make_shared<JoinOperatorFactory<JoinOperator>>();

    add_configuration_if_supported(join_test_configuration);
  }
}

I want it to be discernable which configuration parameter was chosen deliberately and which was not.
```

Reply:
Ok